### DISCUSSION TITLE
Is Use of the Microsoftl.Bcl.AsyncInterfaces Library Actually Supported?

### DISCUSSION THREAD

**Discussion by jimn-purecars** [dotnet/runtime#81352] [Q&A] [answered]
We are interested in using `IAsyncEnumerable<>` in our .NET Framework 4.8 codebase. To do this, I have installed the `Microsft.Bcl.AsyncInterfaces` nuget package, and changed the `<LangVersion>` to `8`. In my test code, things work great!

But according to official [MS docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version), the default language version for Framework 4.8 is `7.3`. In fact, all targets prior to .NET Core 2.1 default to `7.3`. The doc also states:
> Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors

This library is not required from .NET Core 3.0 onwards, so apart from .NET Core 2.1 specifically, it would seem the _only_ use case for this library is in a configuration that is warned against.

So is using e.g. .NET Framework 4.8 with a `<LangVersion>` of `8` actually a supported use case? 

Is the use of this library in this configuration guaranteed to work properly? And is there any guarantee it will continue to work with e.g. new point-releases of .NET Framework 4.8.x?

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by jeffschwMSFT** [comment]
As you pointed out, .NET Framework's supported lang ver is 7.3 (https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version).  This means that moving the lang version ahead may lead to undefined behavior.

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-meta
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

We are interested in using `IAsyncEnumerable<>` in our .NET Framework 4.8 codebase. To do this, I have installed the `Microsft.Bcl.AsyncInterfaces` nuget package, and changed the `<LangVersion>` to `8`. In my test code, things work great!

But according to official [MS docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version), the default language version for Framework 4.8 is `7.3`. In fact, all targets prior to .NET Core 2.1 default to `7.3`. The doc also states:
> Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors

This library is not required from .NET Core 3.0 onwards, so apart from .NET Core 2.1 specifically, it would seem the _only_ use case for this library is in a configuration that is warned against.

So is using e.g. .NET Framework 4.8 with a `<LangVersion>` of `8` actually a supported use case? 

Is the use of this library in this configuration guaranteed to work properly? And is there any guarantee it will continue to work with e.g. new point-releases of .NET Framework 4.8.x?

<table>
  <tr>
    <th align="left">Author:</th>
    <td>jimn-purecars</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`question`, `area-Meta`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by teo-tsirpanis** [comment, marked as answer]
It's a bit complicated situation. First, given that `Microsoft.Bcl.AsyncInterfaces` still receives updates and targets .NET Framework, it should be supported fine. In fact, `IAsyncEnumerable` and `IAsyncDisposable` are just interfaces; there is nothing actual to support in them. They use `ValueTask` from the `System.Tasks.Extensions` package, [which is supported on .NET Framework](https://dotnet.microsoft.com/en-us/platform/support/policy/aspnetcore-2.1).

As for C# 8, it is not officially supported in earlier frameworks, but you can enable it at your own risk. I would do it if I were you and if the codebase is sufficiently tested, but if you don't want to do it, you can use `IAsyncEnumerable` from earlier language versions the manual way:

```csharp
IAsyncEnumerable<int> xs = // …
// C# >= 8
await foreach (int x in xs)
{
    Console.WriteLine(x);
}

// C# < 8
IAsyncEnumerator<int> enumerator = xs.GetAsyncEnumerator();
try
{
    while (await enumerator.MoveNextAsync())
    {
        Console.WriteLine(enumerator.Current);
    }
}
finally
{
    await enumerator.DisposeAsync();
}
```
